Encode using 40, 48 and 56 bits per value

salvatore-campagna opened the pull request:
Note that we use the encoding as follows:
* for values taking [33, 40] bits per value, encode using 40 bits per value
* for values taking [41, 48] bits per value, encode using 48 bits per value
* for values taking [49, 56] bits per value, encode using 56 bits per value

This is an improvement over the encoding used by ForUtils that does not
apply any compression for values taking more than 32 bits per value.

Note that 40, 48 and 56 bits per value represent exact multiples of bytes
(5, 6 and 7 bytes per value). As a result, we always write values using 3, 2
or 1 byte less than the 8 bytes required for a long value.

Looking at the savings in stored bytes, for a block of 128 (long) values
we would normally store 128 x 8 bytes = 1024 bytes, while now we have
the following:
* 40 bits per value: write 645 bytes instead of 1024, saving 379 bytes (37%)
* 48 bits per value: write 772 bytes instead of 1024, saving 252 bytes (24%)
* 56 bits per value: write 897 bytes instead of 1024, saving 127 bytes (12%)

We also apply compression to gauge metrics under the assumption that
compressing values taking more than 32 bits per value works well for
floating point values, because of the way floating point values are
represented (IEEE 754 format).

salvatore-campagna committed:
feature: encode with 40, 48 and 56 bits per value

Note that we use the encoding as follows:
* for values taking [33, 40] bits per value encode using 40 bits per value
* for values taking [41, 48] bits per value encode using 48 bits per value
* for values taking [49, 56] bits per value encode using 56 bits per value

This is an improvement over the encoding used by ForUtils that does
not apply any compression for values taking more than 32 bits per value.

Note that 40, 48 and 56 bits per value represent exact multiples of bytes
(40 bits per value = 5 bytes, 48 bits per value = 6 bytes and 56 bits per
value = 7 bytes). As a result we always write values using 3, 2 or 1 byte
less than the 8 bytes required for a long value.

We also apply compression to gauge metrics under the assumption that
compressing values taking more than 32 bits per value works well for
floating point values, because of the way floating point values
are represented (IEEE 754 format).

salvatore-campagna commented on server/src/main/java/org/elasticsearch/index/codec/tsdb/DocValuesForUtil.java:
```diff
@@ -25,32 +29,79 @@ public DocValuesForUtil(int blockSize) {
+    private void decodeFiveSixOrSevenBytesPerValue(int bitsPerValue, final IndexInput in, long[] out) throws IOException {
+        // NOTE: we expect multibyte values to be written "least significant byte" first
+        for (int longValueIndex = 0; longValueIndex < blockSize; longValueIndex++) {
+            int bytesPerValue = bitsPerValue / Byte.SIZE;
```
Ouch...will move this out of the loop.

salvatore-campagna committed:
fix: use DataInput and DataOutput

salvatore-campagna committed:
fix: remove comment

salvatore-campagna committed:
Update docs/changelog/93371.yaml

elasticsearchmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

elasticsearchmachine commented:
Hi @salvatore-campagna, I've created a changelog YAML for you. *Note* that since this PR is labelled `release highlight`, you need to update the changelog YAML to fill out the extended information sections.

salvatore-campagna committed:
fix: improve readability while skipping bytes

salvatore-campagna committed:
fix: restore non final method params

salvatore-campagna committed:
doc: fix release highlight note

salvatore-campagna committed:
fix: round the number of bits if greater than 32

DocValuesForUtil 'overrides' the number of bits per value
used while encoding a set of values to optimize storage
of values which take more than 32 bits per value. The number
of bits used is written by the encoder and we need to make
sure we write the correct value when serializing it into
the block.

salvatore-campagna commented:
I am runningRally tests locally with `disk-sage-stats` telemetry and will report results after comparing tsdb track test with and without the change.

jpountz commented on server/src/main/java/org/elasticsearch/index/codec/tsdb/DocValuesForUtil.java:
```diff
@@ -25,32 +29,82 @@ public DocValuesForUtil(int blockSize) {
+            for (int byteIndex = 0; byteIndex < bytesPerValue; byteIndex++) {
+                byte b = in.readByte();
+                out[longValueIndex] += ((long) b & 0xFFL) << (8 * byteIndex);
+            }
```
At first sight this decoding logic looks slow as I don't expect it to get vectorized. Did you compare it with something that is more similar to `ForUtil`'s logic?

salvatore-campagna replied:
Maybe if I merge #92934 I can use the microbenchmark here and use those. I don't expect to be vectorized anyway.

salvatore-campagna replied:
@jpountz do you think that a non-vectorized implementation might be too slow (at query time) for us to be able to use it?

jpountz replied:
I'd just like to have something that is as fast as we can make it, given how changing file formats is a bit of a pain in general. So that we're not tempted to go with another layout in 1 month.

salvatore-campagna committed:
fix: save some more bytes not using any padding

salvatore-campagna committed:
fix: read multiple bytes while decoding

salvatore-campagna commented on server/src/main/java/org/elasticsearch/index/codec/tsdb/DocValuesForUtil.java:
```diff
@@ -25,32 +29,82 @@ public DocValuesForUtil(int blockSize) {
+            writer.add(l);
+        }
+        writer.finish();
+    }
```
I can do that and save the padding bytes... even if in the worst case scenario that saves just 3 bytes going from 645 bytes to 642 bytes instead of 1024.

salvatore-campagna commented:
I am working on a vectorized implementation.

StephanErb commented:
> We also apply compression to gauge metrics under the assumption that
> compressing values taking more than 32 bits per value works well for
> floating point values, because of the way floating point values are
> represented (IEEE 754 format).

Have you considered to apply [Gorilla](https://www.vldb.org/pvldb/vol8/p1816-teller.pdf)'s XOR compression on subsequent floats before packing those? This should help to drastically reduce the number of times we'd need to go above 1-2 bytes.

salvatore-campagna commented:
> > We also apply compression to gauge metrics under the assumption that
> > compressing values taking more than 32 bits per value works well for
> > floating point values, because of the way floating point values are
> > represented (IEEE 754 format).
> 
> Have you considered to apply [Gorilla](https://www.vldb.org/pvldb/vol8/p1816-teller.pdf)'s XOR compression on subsequent floats before packing those? This should help to drastically reduce the number of times we'd need to go above 1-2 bytes.

We are already doing that, see `ES87TSDBDocValuesEncoder#forEncode`. In our case it might be even more effective because, before doing xor encoding, we do GCD encoding which might "scale" down the floating point value and later on result in saving more bits after doing the xor operation. Also, we remove the offset if values fall in a range [min, max] by subtracting min and representing the same range as [0, max - min].

jpountz commented on server/src/main/java/org/elasticsearch/index/codec/tsdb/DocValuesForUtil.java:
```diff
@@ -25,32 +29,70 @@ public DocValuesForUtil(int blockSize) {
+    void decode(int bitsPerValue, final DataInput in, long[] out) throws IOException {
         if (bitsPerValue <= 24) {
             forUtil.decode(bitsPerValue, in, out);
         } else if (bitsPerValue <= 32) {
```
should we round numbers of bits between 25 and 32 to 32 and have an equality check here?

jpountz commented on server/src/main/java/org/elasticsearch/index/codec/tsdb/DocValuesForUtil.java:
```diff
@@ -25,32 +29,70 @@ public DocValuesForUtil(int blockSize) {
+    void encode(long[] in, int bitsPerValue, final DataOutput out) throws IOException {
         if (bitsPerValue <= 24) { // these bpvs are handled efficiently by ForUtil
             forUtil.encode(in, bitsPerValue, out);
         } else if (bitsPerValue <= 32) {
```
should we round numbers of bits between 25 and 32 to 32 and have an equality check here?

salvatore-campagna replied:
I think this is not strictly necessary. I think it is more a matter of being consistent between what we write as the `bitsPerValue` and what we do when encoding and decoding. I mean if we write `bitsPerValue = 27` for instance...we encode and decode using 32 bits even if we wrote 27 as `bitsPerValue`. If we do what you suggest we write the actual value 32 and do the encode and decode using 21 bits.

jpountz commented on server/src/main/java/org/elasticsearch/index/codec/tsdb/DocValuesForUtil.java:
```diff
@@ -25,32 +29,70 @@ public DocValuesForUtil(int blockSize) {
+    private void encodeFiveSixOrSevenBytesPerValue(long[] in, int bitsPerValue, final DataOutput out) throws IOException {
+        int bytesPerValue = bitsPerValue / Byte.SIZE;
+        int padding = Long.BYTES - bytesPerValue;
+        byte[] encoded = new byte[bytesPerValue * blockSize + padding];
```
maybe store a byte[1024] as an instance member to avoid allocating a new byte[] every time?

salvatore-campagna commented:
@jpountz I was wondering the following.

We have the ForUtil class that does compression of integer values whose `bitsPerValue` is in the range `[1, 24]`.
Let's forget for a moment that we can extend that range so that the upper bound is larger because that is, maybe,
not worth doing for compression latency/disk save tradeoff reasons.

In case we have `bitsPerValue` > 32 ForUtil does not do anything...as discussed because posting lists will never require integer values larger than 2^32.

Anyway, for values that use more than 32 bits, let's say for instance as an example 55 bitsPerValue we could do the following:
* we have an input array that is long input[128], which is, 128 longs which we can represent using 55 bitsPerValue
* we might split this array in 2, long high[128] and long low[128] which are just the higher 32 bits and lower 32 bits of each input[128]
* now we can apply the ForUtil encoding (and decoding) to both `high` and `low` arrays

My assumption is that the `low` array will be encoded normally using ForUtil on 32 bits, without compression. The `high` array will use less than 32 bits. In our example it will use 55 - 32 = 23 bits.

As a result instead of using 128 * 64 = 8192 bits = 1024 bytes we would use, for instance, 128 x 32 / 8 = 512 bytes to encode the `low` array and 128 x 23 / 8 = 368 bytes to encode the `high` array. We save 1024 - 512 - 368 = 144 bytes which is about 10%.

Observations:
* we do encoding and decoding twice, since we split the original input in two arrays, `high` and `low`, and encode/decode them separately
* we re-use the ForUtil logic for encoding and decoding and its vectorized implementation taking full advantage of it
* the implementation is quite easy since we just need to split the input array if `bitsPerValue` is >32 at encode time and merge the `high` and `low` arrays at decode time.

Does it sound reasonable?

salvatore-campagna commented:
Sorry I messed up the branch pushing the wrong one.

salvatore-campagna committed:
Merge branch 'main' into feature/compress-bitsPerValue-40-48-56

salvatore-campagna committed:
fix: use a buffer and round to 32

salvatore-campagna committed:
fix: make encode and decode public

salvatore-campagna committed:
fix: use NUMERIC_BLOCK_SIZE

salvatore-campagna committed:
fix: test the full range of bitsPerValue

salvatore-campagna commented:
@elasticsearchmachine run elasticsearch-ci/bwc please.

salvatore-campagna committed:
Merge branch 'elastic:main' into feature/compress-bitsPerValue-40-48-56

salvatore-campagna committed:
Merge branch 'elastic:main' into feature/compress-bitsPerValue-40-48-56

salvatore-campagna committed:
Merge branch 'elastic:main' into feature/compress-bitsPerValue-40-48-56

salvatore-campagna commented:
I tried the approach above and as expected the time it takes for encoding doubles because we need to encode/decode two blocks instead of one. As a result I am going to compare just the baseline (what we already have merged, which is no compression for values using more than 32 bits) against what implemented in this PR.

salvatore-campagna committed:
Merge branch 'elastic:main' into feature/compress-bitsPerValue-40-48-56

salvatore-campagna commented:
@martijnvg @jpountz Here is a link to a set of benchmarks I executed to compare encoding/decoding as it is on main now against what we have in this PR.

Keep in mind that the encoding and decoding we have here only kicks in for 40, 48 and 56 bits per value.

https://docs.google.com/spreadsheets/d/1cguyIWZKE6y9Odw53c1ALk2SqlkENdlVsXmVfTgO4R8/edit?usp=sharing

salvatore-campagna commented:
@elasticsearchmachine test this please.

salvatore-campagna committed:
Merge branch 'main' into feature/compress-bitsPerValue-40-48-56

salvatore-campagna committed:
fix: add 20 and 60 bitsPerValue

martijnvg reviewed (approved):
LGTM
(test failures look unrelated to this change)

salvatore-campagna committed:
Merge branch 'elastic:main' into feature/compress-bitsPerValue-40-48-56
